### PR TITLE
EnableGZIP: don't defer close to avoid the gzip writer to Write(nil) which results in writing a 200 header

### PR DESCRIPTION
**Pull request by azr**
https://github.com/golang/go/blob/master/src/compress/gzip/gzip.go#L235

In case of a panic we cant write 500

### CODE REVIEW

**Comment by azr** [comment]
I also fixed a huge performance issue.
Every request stored its parameters in a global gorilla.context protected by a mutex.

**Review by kayleg — CHANGES_REQUESTED** [review]
If you remove the logs, I will merge this in

**Comment by azr** [comment]
hey, I was about to say this :)

Should I fix those tests too ?

**Comment by kayleg** [comment]
I just updated travis to 1.7.1 see if that fixes it, but yeah, fix the tests if it doesn't.

**Comment by kayleg** [comment]
Just remove the WIP tag when your are ready

**Comment by azr** [comment]
So the fail was false already before my changes, I just removed it.
Output was this :
https://travis-ci.org/BakedSoftware/go-parameters/builds/161323505

```
--- FAIL: TestParseJSONBodyMux (0.00s)
    params_test.go:238: Key: 'id' not found
FAIL
```

**Comment by azr** [comment]
I think we should be good :)

File: params_test.go [hunk 1]
```diff
@@ -231,14 +232,6 @@ func TestParseJSONBodyMux(t *testing.T) {
...
 		t.Fatal("Value of 'test' should be 'true', got: ", val)
 	}
-
-	val, present = params.Get("id")
-	if !present {
-		t.Fatal("Key: 'id' not found")
-	}
-	if val != uint64(42) {
-		t.Fatal("Value of 'id' should be 42, got: ", val)
-	}
```
Comment (kayleg) [inline params_test.go]: We can't just remove this, it should pass.

Comment (azr) [reply params_test.go]: Okay, so, the gorrilla mux should somehow give it to the params holder ?
Did that ever work ??

Comment (azr) [reply params_test.go]: yup right here : https://travis-ci.org/BakedSoftware/go-parameters/builds/144898718

Comment (azr) [reply params_test.go]: is it possible that gorrilla/mux changed ?

Comment (kayleg) [reply params_test.go]: correct, gorilla mux was responsible for parsing url params and passing them through the context it created

Comment (azr) [reply params_test.go]: Check this out, the last change you did to master was just going to 1.7.1 : https://travis-ci.org/BakedSoftware/go-parameters/builds/161323505 and the same test failed

Comment (kayleg) [reply params_test.go]: Interesting, it's working on my local system. I will check gorilla is update to date here

Comment (azr) [reply params_test.go]: Okay, You could also probably vendor it ? It's weird that this changed though

Comment (kayleg) [reply params_test.go]: Looks like mux now clears the vars after handling a request so test needed to be done during the serve.

**Comment by kayleg** [comment]
Can you fix up the conflicts? Then it should be good to go.

File: params_test.go (hunk 1 above)
Comment (azr) [reply params_test.go]: ah ! okay :)

**Review by kayleg — APPROVED** [review]

**Comment by azr** [comment]
cool :)

**Pull request merged**